### PR TITLE
Prevent needless copying of image files (BL-2776)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
[Review on Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/769)

### CODE REVIEW

File: src/BloomExe/Edit/EditingView.cs [hunk 1]
```diff
@@ -779,6 +782,29 @@ void SaveChangedImage(GeckoHtmlElement imageElement, PalasoImage imageInfo, stri
...
+		/// creation date, and then the actual bytes in the file, but that's fairly expensive
+		/// for a relatively uncommon occurrence.  We could perhaps use a tool to remove
+		/// unused picture files from a book's folder.
+		/// </remarks>
+		private bool IsSameFile(GeckoHtmlElement imageElement, PalasoImage imageInfo)
+		{
+			var src = imageElement.GetAttribute("src");
+			if (!String.IsNullOrEmpty(src))
+			{
+				var path = Path.Combine(_model.CurrentBook.FolderPath, src.Replace("%20", " "));
```
Comment (hatton) [inline src/BloomExe/Edit/EditingView.cs]: %20 is the obvious one, but presumably what we really need the reverse whatever function put it there. UrlDecode?

File: src/BloomExe/Edit/EditingView.cs [hunk 2]
```diff
@@ -763,6 +763,9 @@ void SaveChangedImage(GeckoHtmlElement imageElement, PalasoImage imageInfo, stri
 			{
 				if (ShouldBailOutBecauseUserAgreedNotToUseJpeg(imageInfo))
 					return;
+				// See https://silbloom.myjetbrains.com/youtrack/issue/BL-2776.
+				if (IsSameFile(imageElement, imageInfo))
```
Comment (hatton) [inline src/BloomExe/Edit/EditingView.cs]: So, I'm not clear how this works. I open ImageToolbox, crop the image, will this preserve the cropping I did, or lose it?

Comment (StephenMcConnel) [inline src/BloomExe/Edit/EditingView.cs]: My original code was not preserving cropping.  The revised code in the method does, while trying to avoid unnecessary duplication of files.

File: src/BloomExe/Edit/EditingView.cs [hunk 3]
```diff
@@ -711,6 +717,7 @@ private void OnChangeImage(DomEventArgs ge)
 				try
 				{
 					imageInfo = PalasoImage.FromFile(existingImagePath);
+					_originalSize = imageInfo.Image.PhysicalDimension;
```
Comment (hatton) [inline src/BloomExe/Edit/EditingView.cs]: OK, so how about if I change... the copyright? What if a future tool lets me change the contrast, etc? I.e. it's not going to work to try and detect that nothing change, short of a byte comparison. I suppose the reason you don't just replace the file is something about replacing the wrong file?

Comment (StephenMcConnel) [inline src/BloomExe/Edit/EditingView.cs]: We had a user whose images for a book all had exactly the same name.  This strikes me as bizarre, but it was a real user reporting that the program was behaving strangely.

I think copyrights are handled okay the way they're working right now, but I probably should double check.  (I remember testing this, but my memory sometimes remembers things that didn't happen.)  A future tool to change contrast, color scheme, etc. would be problematic.

File: src/BloomExe/Edit/EditingView.cs [hunk 4]
```diff
@@ -692,7 +694,8 @@ private void OnChangeImage(DomEventArgs ge)
 			var imageElement = GetImageNode(ge);
 			if (imageElement == null)
 				return;
-			string currentPath = imageElement.GetAttribute("src").Replace("%20", " ");
+			var src = imageElement.GetAttribute ("src");
+			string currentPath = HttpUtilityFromMono.UrlDecode(src);
```
Comment (StephenMcConnel) [inline src/BloomExe/Edit/EditingView.cs]: These changes crept in as part of a previous iteration of this pull request.  Since they're now totally unrelated to the actual bugfix, should they be put in a separate PR as general code cleanup/improvement?

File: src/BloomExe/Edit/PageEditingModel.cs [hunk 5]
```diff
@@ -12,7 +14,8 @@ public class PageEditingModel
 	{
 		public void ChangePicture(string bookFolderPath, GeckoHtmlElement img, PalasoImage imageInfo, IProgress progress)
 		{
-			var imageFileName = ImageUtils.ProcessAndSaveImageIntoFolder(imageInfo, bookFolderPath);
```
Comment (StephenMcConnel) [inline src/BloomExe/Edit/PageEditingModel.cs]: In PageEditingModel.ChangePicture(), a check is now made for whether the chosen picture file is actually the same as the original picture file referenced in the book's HTML.  If it is, then we want to write it without changing its name, whether or not it has actually changed.  If it isn't the same file (even if the base filename is the same), then when it is written, it has to be renamed if the book folder already has a picture file with the same basename.

File: src/BloomExe/ImageProcessing/ImageUtils.cs [hunk 6]
```diff
@@ -51,7 +51,7 @@ Icon    B96B3CB5
 		/// If the image has a filename, replaces any file with the same name.
 		/// </summary>
 		/// <returns>The name of the file, now in the book's folder.</returns>
-		public static string ProcessAndSaveImageIntoFolder(PalasoImage imageInfo, string bookFolderPath)
```
Comment (StephenMcConnel) [inline src/BloomExe/ImageProcessing/ImageUtils.cs]: If the caller tells us that the image file is actually the same file we started out with, then we can build the output path without worrying about overwriting another file.  Otherwise, we have to ensure a unique name for the output file.

**Comment by hatton** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[src/BloomExe/Edit/EditingView.cs, line 699 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzvqx_4BJ-azkXB-olm-r1-699)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/EditingView.cs#L699)):</sup>
I prefer to encourage these little improvements along the way, rather than making it tedious so that we don't do it :-)

---

<sup>**[src/BloomExe/Edit/PageEditingModel.cs, line 25 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzw6BUyh8JjQe60J07j)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/PageEditingModel.cs#L25)):</sup>
Yah, this is awkward, having this much code here in the production file, just for testing.

---

<sup>**[src/BloomExe/Edit/PageEditingModel.cs, line 33 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzw54CJ56huvYgA_6Qu)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/PageEditingModel.cs#L33)):</sup>
This ```isSameFile``` does not appear to be used.

---

<sup>**[src/BloomExe/Edit/PageEditingModel.cs, line 50 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzw4dq2tS-K9dLr_wW-)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/PageEditingModel.cs#L50)):</sup>
As far as I can tell, this method wants to be named something like "IsSameFilePath()". The current name harkens back to the previous attempt to actually know that the file has not changed in some way other than comparing file contents.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769)

**Comment by StephenMcConnel** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[src/BloomExe/Edit/PageEditingModel.cs, line 33 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzw54CJ56huvYgA_6Qu)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/PageEditingModel.cs#L33)):</sup>
Done.

---

<sup>**[src/BloomExe/Edit/PageEditingModel.cs, line 50 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769#-Jzw4dq2tS-K9dLr_wW-)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ce0770d5051a8ed7ce98bf25e075441b9325fda1/src/BloomExe/Edit/PageEditingModel.cs#L50)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/769)
